integration/querier_remote_read_test.go

```diff
@@ -59,164 +63,258 @@ func TestQuerierRemoteRead(t *testing.T) {
 		},
-		"float histograms": {
-			expectedValType: chunkenc.ValFloatHistogram,
-			histograms: func(startMs, endMs int64) []prompb.Histogram {
-				var hists []prompb.Histogram
-				for i := startMs; i < endMs; i++ {
-					h := test.GenerateTestFloatHistogram(int(i))
-					hists = append(hists, remote.FloatHistogramToHistogramProto(i, h))
-				}
-				return hists
+		"float samples, with hints time range different than query time range": {
+			valType:    chunkenc.ValFloat,
+			metricName: floatMetricName,
+			query: &prompb.Query{
+				Matchers:         remoteReadQueryMatchersByMetricName(floatMetricName),
+				StartTimestampMs: pushedStartMs,
+				EndTimestampMs:   pushedEndMs,
+				Hints: &prompb.ReadHints{
+					StartMs: now.UnixMilli(),
+					EndMs:   now.Add(10 * time.Second).UnixMilli(),
+				},
 			},
+			// Mimir doesn't cut returned chunks to the requested time range for performance reasons. This means
+			// that we get the entire chunks in output. TSDB targets to cut chunks once every 120 samples, but
+			// it's based on estimation and math is not super accurate. That's why we get these not-perfectly-aligned
+			// chunk ranges. However, they're consistent, so tests are stable.
+			expectedStartMs: now.Add(-1500 * time.Millisecond).UnixMilli(),
+			expectedEndMs:   now.Add(10100 * time.Millisecond).UnixMilli(),
+		},
+		"histograms": {
+			valType:         chunkenc.ValHistogram,
+			metricName:      histogramMetricName,
+			query:           remoteReadQueryByMetricName(histogramMetricName, time.UnixMilli(pushedStartMs), time.UnixMilli(pushedEndMs)),
+			expectedStartMs: pushedStartMs,
+			expectedEndMs:   pushedEndMs,
+		},
+		"float histograms": {
+			valType:         chunkenc.ValFloatHistogram,
+			metricName:      floatHistogramMetricName,
+			query:           remoteReadQueryByMetricName(floatHistogramMetricName, time.UnixMilli(pushedStartMs), time.UnixMilli(pushedEndMs)),
+			expectedStartMs: pushedStartMs,
+			expectedEndMs:   pushedEndMs,
 		},
 	}
 
-	for testName, tc := range testCases {
-		t.Run(testName, func(t *testing.T) {
-			s, err := e2e.NewScenario(networkName)
-			require.NoError(t, err)
-			defer s.Close()
+	s, err := e2e.NewScenario(networkName)
+	require.NoError(t, err)
+	defer s.Close()
 
-			flags := mergeFlags(BlocksStorageFlags(), BlocksStorageS3Flags(), map[string]string{
-				"-distributor.ingestion-rate-limit": "1048576",
-				"-distributor.ingestion-burst-size": "1048576",
-				"-distributor.remote-timeout":       "10s",
-			})
+	flags := mergeFlags(BlocksStorageFlags(), BlocksStorageS3Flags(), map[string]string{
+		"-distributor.ingestion-rate-limit": "1048576",
+		"-distributor.ingestion-burst-size": "1048576",
+		"-distributor.remote-timeout":       "10s",
 
-			// Start dependencies.
-			minio := e2edb.NewMinio(9000, blocksBucketName)
+		// This test writes samples sparse in time. We don't want compaction to trigger while testing.
+		"-blocks-storage.tsdb.block-ranges-period": "2h",
```

Comment:
Note to reviewers: the default in integration tests is 1m, but I don't want such a small block period range in these tests. Reason is that chunks are also cut to period ranges, making the assertion even more complicated.